In [1]:
import scanpy as sc
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

In [2]:
import scanpy as sc
from pathlib import Path

raw_data_path = Path("data/raw/tabula_muris_senis_dev.h5ad")

print("File exists:", raw_data_path.exists())

File exists: False


## Why this notebook exists

The first two notebooks used PBMC3k only to learn the Scanpy workflow. This notebook begins the actual aging part of the project by loading Tabula Muris Senis and checking whether the dataset contains age, tissue, mouse, and cell-type metadata.

I chose Tabula Muris Senis because it is a public mouse aging single cell atlas with age, tissue, mouse, and cell type metadata. This makes it suitable for testing whether age related transcriptional patterns can be detected at single cell resolution.

In [3]:
import os

print("Current notebook folder:")
print(os.getcwd())

print("\nFiles here:")
print(os.listdir())

Current notebook folder:
/Users/palaksharma/aging-clock-env

Files here:
['bin', 'include', '03_tabula_muris_subset_preparation.ipynb.ipynb', 'etc', 'pyvenv.cfg', '.gitignore', 'lib', '.ipynb_checkpoints', '02_data_analysis.ipynb', '01_data_loading_and_QC.ipynb', 'data', 'share']


In [4]:
print("Does data folder exist?", os.path.exists("data"))
print("Does raw folder exist?", os.path.exists("data/raw"))

if os.path.exists("data/raw"):
    print("Files inside data/raw:")
    print(os.listdir("data/raw"))

Does data folder exist? True
Does raw folder exist? False


In [5]:
import os

print(os.getcwd())

/Users/palaksharma/aging-clock-env


In [6]:
from pathlib import Path

raw_data_path = Path("/Users/palaksharma/single-cell-aging-clock/data/raw/tabula_muris_senis_dev.h5ad")

print("File exists:", raw_data_path.exists())

File exists: True


In [7]:
import scanpy as sc

aging_data_preview = sc.read_h5ad(raw_data_path, backed="r")

print(aging_data_preview)

AnnData object with n_obs × n_vars = 5218 × 22966 backed at '/Users/palaksharma/single-cell-aging-clock/data/raw/tabula_muris_senis_dev.h5ad'
    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden'
    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'age_colors', 'cell_ontology_class_colors', 'leiden', 'louvain', 'neighbors', 'pca'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'


/Users/palaksharma/aging-clock-env/lib/python3.14/site-packages/anndata/_io/h5ad.py:192: FutureWarning: Moving element from .uns['neighbors']['distances'] to .obsp['distances'].

This is where adjacency matrices should go now.
  adata = AnnData(**d)
/Users/palaksharma/aging-clock-env/lib/python3.14/site-packages/anndata/_io/h5ad.py:192: FutureWarning: Moving element from .uns['neighbors']['connectivities'] to .obsp['connectivities'].

This is where adjacency matrices should go now.
  adata = AnnData(**d)


In [8]:
import scanpy as sc
from pathlib import Path

raw_data_path = Path("/Users/palaksharma/single-cell-aging-clock/data/raw/tabula_muris_senis_dev.h5ad")

print("File exists:", raw_data_path.exists())

aging_data_preview = sc.read_h5ad(raw_data_path, backed="r")

print(aging_data_preview)
print(aging_data_preview.obs.columns.tolist())

File exists: True
AnnData object with n_obs × n_vars = 5218 × 22966 backed at '/Users/palaksharma/single-cell-aging-clock/data/raw/tabula_muris_senis_dev.h5ad'
    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden'
    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'age_colors', 'cell_ontology_class_colors', 'leiden', 'louvain', 'neighbors', 'pca'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
['FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden']


/Users/palaksharma/aging-clock-env/lib/python3.14/site-packages/anndata/_io/h5ad.py:192: FutureWarning: Moving element from .uns['neighbors']['distances'] to .obsp['distances'].

This is where adjacency matrices should go now.
  adata = AnnData(**d)
/Users/palaksharma/aging-clock-env/lib/python3.14/site-packages/anndata/_io/h5ad.py:192: FutureWarning: Moving element from .uns['neighbors']['connectivities'] to .obsp['connectivities'].

This is where adjacency matrices should go now.
  adata = AnnData(**d)


In [9]:
print(aging_data_preview)

AnnData object with n_obs × n_vars = 5218 × 22966 backed at '/Users/palaksharma/single-cell-aging-clock/data/raw/tabula_muris_senis_dev.h5ad'
    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden'
    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'age_colors', 'cell_ontology_class_colors', 'leiden', 'louvain', 'neighbors', 'pca'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'


In [10]:
metadata_columns = aging_data_preview.obs.columns.tolist()

print("Number of metadata columns:", len(metadata_columns))
print(metadata_columns)

Number of metadata columns: 15
['FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden']


In [12]:
possible_age_columns = []
possible_tissue_columns = []
possible_celltype_columns = []

for column in metadata_columns:
    col_lower = column.lower()
    
    if "age" in col_lower:
        possible_age_columns.append(column)
        
    if "tissue" in col_lower or "organ" in col_lower:
        possible_tissue_columns.append(column)
        
    if "cell" in col_lower or "type" in col_lower or "annotation" in col_lower:
        possible_celltype_columns.append(column)

print("Found these options for Age:", possible_age_columns)
print("Found these options for Tissue/Organ:", possible_tissue_columns)
print("Found these options for Cell-Type:", possible_celltype_columns)

Found these options for Age: ['age']
Found these options for Tissue/Organ: ['subtissue', 'tissue']
Found these options for Cell-Type: ['cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation']


In [13]:
age_column = "age"

print(aging_data_preview.obs[age_column].value_counts())

age
18m    1960
24m    1877
3m     1381
Name: count, dtype: int64


In [14]:
tissue_column = "tissue"

print(aging_data_preview.obs[tissue_column].value_counts().head(20))

tissue
Lung    5218
Name: count, dtype: int64


In [15]:
cell_type_column = "cell_ontology_class"

print(aging_data_preview.obs[cell_type_column].value_counts().head(20))

cell_ontology_class
bronchial smooth muscle cell                       2117
fibroblast of lung                                  483
myeloid dendritic cell                              373
adventitial cell                                    289
B cell                                              276
classical monocyte                                  202
type II pneumocyte                                  167
vein endothelial cell                               152
non-classical monocyte                              149
neutrophil                                          123
CD8-positive, alpha-beta T cell                     119
smooth muscle cell of the pulmonary artery          111
endothelial cell of lymphatic vessel                 99
dendritic cell                                       62
CD4-positive, alpha-beta T cell                      59
NK cell                                              52
ciliated columnar cell of tracheobronchial tree      52
pulmonary interstitial fibro

In [16]:
print(aging_data_preview.obs["free_annotation"].value_counts().head(20))

free_annotation
Capillary                     1790
Alveolar Fibroblast            483
Adventitial Fibroblast         289
Capillary Aerocyte             280
Myeloid Dendritic Type 2       239
Classical Monocyte             202
B                              188
Alveolar Epithelial Type 2     164
Vein                           152
Nonclassical Monocyte          149
Myeloid Dendritic Type 1       134
Neutrophil                     123
CD8+ T                         119
Artery                         111
Lympatic                        98
Zbtb32+ B                       82
CD4+ T                          59
Ccr7+ Dendritic                 57
Natural Killer                  52
Myofibroblast                   50
Name: count, dtype: int64


In [19]:

metadata_table = aging_data_preview.obs.copy()
selected_cell_ids = []
cells_per_age_group = 2000

unique_ages = metadata_table["age"].unique()
print(f" Found these age groups to process: {unique_ages}\n")

for age in unique_ages:
    cells_in_this_age = metadata_table[metadata_table["age"] == age]
    
    total_available_cells = len(cells_in_this_age)
    print(f" Age {age}: Total available cells = {total_available_cells}")
    
    if total_available_cells > cells_per_age_group:
        number_to_take = cells_per_age_group
    else:
        number_to_take = total_available_cells
        
    print(f"   -> Sampling {number_to_take} cells from this group")
    
    sampled_subset = cells_in_this_age.sample(n=number_to_take, random_state=42)
    
    cell_ids_list = sampled_subset.index.tolist()
    selected_cell_ids.extend(cell_ids_list)


print(f" Total selected cells for our model: {len(selected_cell_ids)}")


 Found these age groups to process: ['18m', '24m', '3m']
Categories (3, object): ['3m', '18m', '24m']

 Age 18m: Total available cells = 1960
   -> Sampling 1960 cells from this group
 Age 24m: Total available cells = 1877
   -> Sampling 1877 cells from this group
 Age 3m: Total available cells = 1381
   -> Sampling 1381 cells from this group
 Total selected cells for our model: 5218


In [20]:
aging_subset = aging_data_preview[selected_cell_ids, :].to_memory()

print(aging_subset)

AnnData object with n_obs × n_vars = 5218 × 22966
    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden'
    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'age_colors', 'cell_ontology_class_colors', 'leiden', 'louvain', 'neighbors', 'pca'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'


In [21]:
print(aging_subset.obs["age"].value_counts())

age
18m    1960
24m    1877
3m     1381
Name: count, dtype: int64


In [22]:
print(aging_subset.obs["tissue"].value_counts().head(20))

tissue
Lung    5218
Name: count, dtype: int64


In [23]:
print(aging_subset.obs["cell_ontology_class"].value_counts().head(20))

cell_ontology_class
bronchial smooth muscle cell                       2117
fibroblast of lung                                  483
myeloid dendritic cell                              373
adventitial cell                                    289
B cell                                              276
classical monocyte                                  202
type II pneumocyte                                  167
vein endothelial cell                               152
non-classical monocyte                              149
neutrophil                                          123
CD8-positive, alpha-beta T cell                     119
smooth muscle cell of the pulmonary artery          111
endothelial cell of lymphatic vessel                 99
dendritic cell                                       62
CD4-positive, alpha-beta T cell                      59
NK cell                                              52
ciliated columnar cell of tracheobronchial tree      52
pulmonary interstitial fibro

In [25]:
from pathlib import Path

processed_subset_path = Path("/Users/palaksharma/single-cell-aging-clock/data/processed/tabula_muris_senis_10k_subset.h5ad")

aging_subset.write(processed_subset_path)

print("Saved subset here:")
print(processed_subset_path)

Saved subset here:
/Users/palaksharma/single-cell-aging-clock/data/processed/tabula_muris_senis_10k_subset.h5ad


In [26]:
print(aging_data_preview.obs["age"].value_counts())

age
18m    1960
24m    1877
3m     1381
Name: count, dtype: int64


In [27]:
age_to_months = {
    "3m": 3,
    "18m": 18,
    "24m": 24
}

aging_data_preview.obs["age_months"] = aging_data_preview.obs["age"].map(age_to_months)

print(aging_data_preview.obs[["age", "age_months"]].head())

                                       age age_months
index                                                
A10_B000229_B009899_S10.mm10-plus-4-0  18m         18
A10_B000248_B009895_S10.mm10-plus-4-0  18m         18
A10_B000265_B009901_S10.mm10-plus-4-0  18m         18
A10_B000266_B009900_S10.mm10-plus-4-0  18m         18
A10_B002258_B009940_S10.mm10-plus-4-0  18m         18


In [28]:
aging_subset = aging_data_preview.to_memory()

print(aging_subset)

AnnData object with n_obs × n_vars = 5218 × 22966
    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden', 'age_months'
    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'age_colors', 'cell_ontology_class_colors', 'leiden', 'louvain', 'neighbors', 'pca'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'


In [29]:
aging_subset.obs["age_months"] = aging_subset.obs["age"].map(age_to_months)

print(aging_subset.obs["age_months"].value_counts().sort_index())

age_months
3     1381
18    1960
24    1877
Name: count, dtype: int64


In [30]:
print(aging_subset.obs["tissue"].value_counts())

tissue
Lung    5218
Name: count, dtype: int64


In [31]:
print(aging_subset.obs["cell_ontology_class"].value_counts().head(20))

cell_ontology_class
bronchial smooth muscle cell                       2117
fibroblast of lung                                  483
myeloid dendritic cell                              373
adventitial cell                                    289
B cell                                              276
classical monocyte                                  202
type II pneumocyte                                  167
vein endothelial cell                               152
non-classical monocyte                              149
neutrophil                                          123
CD8-positive, alpha-beta T cell                     119
smooth muscle cell of the pulmonary artery          111
endothelial cell of lymphatic vessel                 99
dendritic cell                                       62
CD4-positive, alpha-beta T cell                      59
NK cell                                              52
ciliated columnar cell of tracheobronchial tree      52
pulmonary interstitial fibro

In [32]:
from pathlib import Path

processed_subset_path = Path(
    "/Users/palaksharma/single-cell-aging-clock/data/processed/tabula_muris_senis_dev_subset.h5ad"
)

aging_subset.write(processed_subset_path)

print("Saved aging development dataset:")
print(processed_subset_path)

Saved aging development dataset:
/Users/palaksharma/single-cell-aging-clock/data/processed/tabula_muris_senis_dev_subset.h5ad


The dataset contains 5,218 cells across three age groups:

- 3 months: 1,381 cells
- 18 months: 1,960 cells
- 24 months: 1,877 cells

Since the file is already small enough for local development, I kept all cells instead of subsampling.